File: modules/day-04-reliable/python/solution.py
def pass_rate(scores, passing_score=60):
    if not scores:
        raise ValueError("scores must contain at least one value")

    passed = [score >= passing_score for score in scores]
    return sum(passed) / len(scores) * 100


def run_checks():
    assert pass_rate([55, 60, 62, 90]) == 75
    assert pass_rate([60, 60, 60]) == 100

    try:
        pass_rate([])
    except ValueError:
        pass
    else:
        raise AssertionError("Empty input should raise ValueError")

    print("All checks passed")


if __name__ == "__main__":
    run_checks()

File: modules/day-04-reliable/r/solution.R
pass_rate <- function(scores, passing_score = 60) {
  if (length(scores) == 0) {
    stop("scores must contain at least one value")
  }

  passed <- scores >= passing_score
  sum(passed) / length(scores) * 100
}

run_checks <- function() {
  stopifnot(pass_rate(c(55, 60, 62, 90)) == 75)
  stopifnot(pass_rate(c(60, 60, 60)) == 100)

  empty_input_failed <- FALSE

  tryCatch(
    {
      pass_rate(c())
    },
    error = function(e) {
      empty_input_failed <<- TRUE
    }
  )

  stopifnot(empty_input_failed)
  cat("All checks passed\n")
}

run_checks()
